webpack/assets/javascripts/react_app/components/ForemanModal/ForemanModalReducer.js

```diff
@@ -0,0 +1,56 @@
+import Immutable from 'seamless-immutable';
+import { translate as __ } from '../../../react_app/common/I18n';
+import {
+  SET_MODAL_OPEN,
+  SET_MODAL_CLOSED,
+  ADD_MODAL,
+} from './ForemanModalConstants';
+
+const initialState = Immutable({});
+
+// Modals state has id as key and open state as value:
+// { myModal: {open: true} }
+// Since keys cannot be duplicated, we avoid creating duplicate modals in this way.
+
+export default (state = initialState, action) => {
+  switch (action.type) {
+    case SET_MODAL_OPEN: {
+      if (!state[action.payload.id]) {
+        throw new Error(
+          __(
+            `SET_MODAL_OPEN error: Modal with id '${action.payload.id}' does not exist`
+          )
+        );
+      }
+      const newState = {};
+      newState[action.payload.id] = {
+        open: true,
+      };
+      return state.merge(newState);
+    }
+    case SET_MODAL_CLOSED: {
+      if (!state[action.payload.id]) {
+        throw new Error(
+          __(
+            `SET_MODAL_CLOSED error: Modal with id '${action.payload.id}' does not exist`
+          )
+        );
+      }
+      const newState = {};
+      newState[action.payload.id] = {
+        open: false,
+      };
+      return state.merge(newState);
+    }
+    case ADD_MODAL: {
+      if (state[action.payload.id]) return state; // if it already exists, don't change its state
```

Comment:
can we filter this in the action? This returns the same state but all the middlewares were activated for nothing

Reply:
@amirfefer true, will update!

Reply:
Update: I realized that in order to do this, we will need access to the store state from within the action.  This would require one of the following:

1. In `ForemanModalActions.js`, import the Redux store directly and use `store.getState()`
2. Make all of the ForemanModal actions into thunks and use the `getState` provided by `redux-thunk`
3. Pass the store state into the action creator as a function parameter.

I don't like any of these.  On the other hand, going through our middleware shouldn't really cause any noticeable performance issue because it would be filtered out by all the middleware logic and passed straight through to the reducer.

Reply:
while redux-thunk's `getState` is simple and has a constant cost, middlewares stack tends to grow, plus I'd prefer having cleaner reducers by moving logic (i.e side effects, filtering, error handling...) to actions.

Reply:
@amirfefer I agree that of the 3 options, `getState` seems to be the lesser of the evils.  Stay tuned for update

Reply:
any update? :)

Reply:
The action creator now handles the error message, but I think I left that line in so that you can't set a modal open or closed with `ADD_MODAL`.  (You could theoretically still dispatch the plain action without using the action creator.)